### ISSUE TITLE
IIIF scale_factors and maxLevel

### ISSUE THREAD

**Issue by hyperknot** [openseadragon/openseadragon#1394]
I've been going through understanding different IIIF viewer implementations (Leaflet-IIIF, Klokantech-OpenLayers and OSM), and I have to say that OSD clearly seems like a reference implementation, it's way higher quality above the others.

Having said that, there is one point, which I believe is a bug.
It is [this line](https://github.com/openseadragon/openseadragon/blob/01408a47445278e888d9f646b93f7aa910382122/src/iiiftilesource.js#L130), the second maxLevel in the snippet below: 

    if (!options.maxLevel && !this.emulateLegacyImagePyramid) {
        if (!this.scale_factors) {
            options.maxLevel = Number(Math.ceil(Math.log(Math.max(this.width, this.height), 2)));
        } else {
            options.maxLevel = Math.floor(Math.pow(Math.max.apply(null, this.scale_factors), 0.5));
        }
    }

From the IIIF specs, I believe scaleFactors means what reduction levels are present. Thus
[1] -> only 100% resolution tiles are present
[1, 2] -> 50% and 100% resolution tiles
[1, 2, 4] -> 25%, 50% and 100% resolution tiles are present

Now if you calculate, the above function will calculate:
[1] -> 1
[1, 2] -> 1
[1, 2, 4] -> 2

Now, it's clear that [1] and [1, 2] shouldn't be the same. The right value for [1] is actually 0.

This can be tested on tweaked static file IIIF sources, like:
https://github.com/zimeon/iiif/tree/master/demo-static

I can share a .zip if you'd like to reproduce the bug.

**Comment by hyperknot** [comment]
I believe the correct value should be simply the log2 of scale_factor, like 

```
Math.log(scale_factor) * Math.LOG2E
```

I see that there is a demo folder for test cases. If you'd like I can make a new one from `test/data/iiif_2_0_tiled` which reproduces this bug.

**Comment by iangilman** [comment]
@hyperknot thank you for discovering this!

Any thoughts, @jpstroop @azaroth42?

**Comment by azaroth42** [comment]
I agree with @hyperknot. I suspect that line was simply trying to reverse out the previous math, rather than thinking about it with an open mind. 

:+1: to the Math.LOG2E solution, and thank you for finding the bug!

**Comment by hyperknot** [comment]
In the meanwhile I found out we need `Math.round` both for `* Math.LOG2E` and `/ Math.log(2)` based solutions for some values.

So I recommend the following solution:

```
var maxScaleFactor = Math.max.apply(null, this.scale_factors)
options.maxLevel = Math.round(Math.log(maxScaleFactor) * Math.LOG2E)
```

**Comment by iangilman** [comment]
@hyperknot Sounds good... can you make a pull request with the fix? 

@azaroth42 Thanks for taking a look!

**Comment by hyperknot** [comment]
@iangilman sure, I'll do!

**Comment by hyperknot** [comment]
Made PR: #1401 

Also added test based on the iiif_2_0_tiled one.

**Comment by iangilman** [comment]
Wonderful... I appreciate the extra test :)

**Comment by iangilman** [comment]
Fixed by #1401
